net: Namespace network symbols to avoid conflicts with POSIX or libc

jukkar opened the pull request:
Rename network symbols i.e., add `net_`, `NET_` or `ZSOCK_` prefixes to those network symbols that can be found in POSIX or libc. This way we can avoid circular dependency issues.

Fixes #97050

jukkar committed:
net: Namespace network symbols to avoid conflicts with Posix/libc

Rename network symbols i.e., add net_, NET_ or ZSOCK_ prefixes
to those network symbols that can be found in Posix or libc.
This way we can avoid circular dependency issues.

Add also a compatibility header file that allows user to continue
use various network APIs without adding either net_, NET_ or ZSOCK_
prefixes.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Convert Ethernet drivers to use renamed network APIs

Rename network symbols in Ethernet drivers to use the renamed network APIs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Convert wifi drivers to use renamed network APIs

Rename network symbols in wifi drivers to use the renamed
network APIs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Convert usb drivers to use renamed network APIs

Rename network symbols in usb drivers to use the renamed network APIs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Convert network codebase to use renamed network APIs

Rename network symbols in network stack to use the renamed network APIs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Convert modem drivers to use renamed network APIs

Rename network symbols in modem drivers to use the renamed
network APIs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Convert IEEE 802.15.4 drivers to use renamed network APIs

Rename network symbols in IEEE 802.15.4 drivers to use the renamed
network APIs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Convert network drivers to use renamed network APIs

Rename network symbols in net drivers to use the renamed network APIs.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: Network should not include any Posix header

Removing Posix header inclusion as networking APIs and code should
be self contained now.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: sockets: Remove Posix header file includes

Do not include Posix header files inside network stack as we should
not depend on Posix symbols in net stack.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
lib: posix: net: Add support for namespaced network symbols

Use namespaced network symbols in order to avoid circular dependency
between Posix and network subsystems.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

cfriedt commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -656,9 +656,9 @@ static sa_family_t eth_intel_igc_get_sa_family(uint8_t *rx_buf)
 
 	switch (ntohs(eth_hdr->type)) {
```
we should have a non-POSIX version of `ntohs()` and others too.

jukkar replied:
Yes, I will change these. I started to change the source all over the code base but there was just so much of it, so I decided to concentrate my efforts first to subsys/net, tests/net and include/zephyr/net directories. Some of the drivers contain already changes and I will continue this work.

cfriedt reviewed (commented):
Good work so far. I just left a couple of comments for non-sockaddr things. Although I'm sure you are already aware and planning for those too.

jukkar commented on doc/connectivity/networking/api/coap_server.rst:
```diff
@@ -142,7 +142,7 @@ of CoAP services. An example using a temperature sensor can look like:
 
     static int send_temperature(struct coap_resource *resource,
-                                const struct sockaddr *addr, socklen_t addr_len,
+                                const struct net_sockaddr *addr, socklen_t addr_len,
```
I did not go that route, but just changed `socklen_t` to `net_socklen_t`. Not sure if we should do that `uint32_t` change as now it is easy to map the socklen variable between the network and Posix apis (because the naming is similar).
What other reviewers think about this, I am open to replace `net_socklen_t` if there are good arguments for it.

aescolar commented on include/zephyr/net/net_compat.h:
```diff
@@ -0,0 +1,303 @@
+#ifndef ZEPHYR_INCLUDE_NET_NET_COMPAT_H_
+#define ZEPHYR_INCLUDE_NET_NET_COMPAT_H_
+
+#if defined(CONFIG_NET_NAMESPACE_COMPAT_MODE)
```
I wonder should this ifdef be in include/zephyr/net/net_ip.h instead, and 
should `CONFIG_NET_NAMESPACE_COMPAT_MODE` start being `default y` always?

Then The POSIX_API can include this header on its own from wherever it needs, and CONFIG_NET_NAMESPACE_COMPAT_MODE be switched to `default n` and deprecated(?)

aescolar reviewed (commented):
Thanks @jukkar , overall it is looking good to me :)

jukkar committed:
samples: net: Add relevant Posix headers for networking

We need certain network related Posix header files so that the
network sample compiles ok. Previously the Posix symbols were
introduced by network stack automatically even if user had not
enabled Posix, but that is no longer the case.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: Add relevant Posix headers for networking

We need certain network related Posix header files so that the
network tests using Posix symbols compiles ok.
Previously the Posix symbols were introduced by network stack
automatically, but that is no longer the case.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: socket: tls_configurations: err variable might be unused

Mark the err variable as possibly unsed depending on the enabled
config options.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: zperf: Change old code to pass compliance checker

Compliance checker complains

   TYPO_SPELLING: 'in in' may be misspelled - perhaps 'is in'?

for code like this

   static struct net_sockaddr_in in_addr_my = {

so change the variable to ipv4_addr_my. Similar change is
done for IPv6 variables for consistency.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/l2/virtual/ipip/ipip.c:
```diff
@@ -191,12 +191,12 @@ static int interface_send(struct net_if *iface, struct net_pkt *pkt)
 		tmp = net_pkt_alloc_with_buffer(iface,
 						sizeof(struct net_ipv4_hdr),
-						AF_INET, IPPROTO_IP,
+						NET_AF_INET, IPPROTO_IP,
```
Same story with `IPPROTO_IP`.

rlubos reviewed (commented):
Found some leftover symbols, we probably need some test variant that builds the net stack w/o the compatibility mode enabled to catch those.

jukkar committed:
log: backend: net: Remove Posix dependency

The network backend needs nothing from Posix so remove the
dependency.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: all: Add more compilation support

Make sure we test coap client/server, MQTT SN, telnet and
logging network backend when testing all network options.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
modules: openthread: Remove SOCK_NONBLOCK as it is not supported

There is no SOCK_NONBLOCK support in zephyr so remove the option.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented:
> Ping to reviewers, please help to review this PR. As the code changes lot of files that use networking APIs, it is a constant race with other PRs and it would be nice to get in sooner than later.

I'm sorry to question this, but this will break a lot out-of-tree users, no? Shouldn't we have a transition in place? This would allow to split the in-tree users into multiple smaller PRs.

EDIT: I didn't see the compat header yet, my browser has difficulties opening the PR.

jukkar commented:
Accidentally added Treewide label although this PR is only touching network related API usage, removed the label.

jukkar commented:
> I think the one giant commit touching everything is bad. A better approach IMO would be to add namespacing to definitions/structs/symbols **and** add the compatibility header in a single (small) commit.
> And introduce changes to other parts of the tree in separate commits.

What I can / plan to do here
* Move the net_ip.h and socket.h changes to a separate smallish commit (for easier review of the API changes)
* I am not planning to split the big commit that changes the API usage in zephyr. It was a quite big task to even create this one commit and splitting it to smaller pieces requires so much work I am not comfortable to do.
* Add doc to migration guide about the changes and what is required for the applications

> I'm sorry to question this, but this will break a lot out-of-tree users, no?

There is a compatibility flag that is currently set to "y" by default. This means that no changes should needed in external applications. If the external application uses Posix symbols like `ntohs()` etc. and if the compatibility flag is "n", then it needs to include relevant Posix include files as just including either `net_ip.h` or `socket.h` from net will not provide those symbols any more. Other than that, no changes would be needed in application side. The compatibility layer is already there as you noticed and should provide backward compatibility to the apps so we can do things in a controlled way.

jukkar commented:
The latest version has these commits, I have description of each one and how we could split them further:

* net: Namespace network symbols to avoid conflicts with Posix/libc
    Rename network symbols i.e., add net_, NET_ or ZSOCK_ prefixes
    to those network symbols that can be found in Posix or libc.
    This way we can avoid circular dependency issues.

* net: Add network namespace compatibility header file
    Allow user to continue use various network APIs without adding
    either net_, NET_ or ZSOCK_ prefixes.

The previous two can be applied separately if needed. I tested that
twister run will pass ok if `CONFIG_NET_NAMESPACE_COMPAT_MODE=y`
which is the default.

* modules: openthread: Remove SOCK_NONBLOCK as it is not supported
    There is no SOCK_NONBLOCK support in zephyr so remove the option.
    This is sent separately as #99371

* net: Convert Zephyr codebase to use renamed network APIs
    Rename network symbols in Zephyr to use the renamed network APIs.
    This is the main meat of the PR that changes all the network
    symbols in Zephyr code. This can be split if needed, perhaps
    separating the core network stack, network tests and samples in
    this PR or in a separate PR, and then and the rest in either
    one or multiple PRs.

The rest could be applied after the all the previous PRs are merged.

* net: socket: Network should not include any Posix header
    Removing Posix header inclusion as networking APIs and code should
    be self contained now.

* net: sockets: Remove Posix header file includes
    Do not include Posix header files inside network stack as we should
    not depend on Posix symbols in net stack.

* lib: posix: net: Add support for namespaced network symbols
    Use namespaced network symbols in order to avoid circular dependency

* samples: net: Add relevant Posix headers for networking
    We need certain network related Posix header files so that the
    network sample compiles ok. Previously the Posix symbols were
    introduced by network stack automatically even if user had not
    enabled Posix, but that is no longer the case.

* tests: net: Add relevant Posix headers for networking
    We need certain network related Posix header files so that the
    network tests using Posix symbols compiles ok.
    Previously the Posix symbols were introduced by network stack
    automatically, but that is no longer the case.

* tests: net: socket: tls_configurations: err variable might be unused
    Mark the err variable as possibly unsed depending on the enabled
    config options.

* net: zperf: Change old code to pass compliance checker
    Compliance checker fixes to zperf.

* log: backend: net: Remove Posix dependency
    The network backend needs nothing from Posix so remove the
    dependency.

* tests: net: all: Add more compilation support
    Make sure we test coap client/server, MQTT SN, telnet and
    logging network backend when testing all network options.

Any proposal how to proceed? The main problem is the 4th commit which does the big switch. Personally I would do the core network changes in one big commit together with network tests and samples. The drivers etc misc stuff could be separated from this so that relevant maintainers could review them more easily.

jukkar commented:
> Isn't this what `coccinelle` can be used for? If so, we could provide a script for downstream migration too.
> _disclaimer: this is a guess, I have no experience with it_

Sure, I know the tool but have never used it, just have heard cursing from someone who created such script.

rlubos reviewed (dismissed):
LGTM, IMHO I don't see a strong need to split commits up. Whatever we do the change-set will be enormous anyway with such a major overhaul.

pdgendt reviewed (dismissed):
> LGTM, IMHO I don't see a strong need to split commits up. Whatever we do the change-set will be enormous anyway with such a major overhaul.

Don't want to block either, let's flip the switch.

pdgendt commented:
Needs a rebase for the compliance issue to be fixed.

maass-hamburg commented on subsys/logging/backends/Kconfig.net:
```diff
@@ -6,7 +6,6 @@
 config LOG_BACKEND_NET
 	bool "Networking backend"
 	depends on NETWORKING && (NET_UDP || NET_TCP) && NET_SOCKETS && !LOG_MODE_IMMEDIATE
-	depends on POSIX_C_LANG_SUPPORT_R
```
it does https://github.com/zephyrproject-rtos/zephyr/blob/39065c6ee6eb2874849c675e06c74094c962a1d7/subsys/logging/log_output.c#L188-L193
